rust/sbp/src/sbp2json.rs

```diff
@@ -0,0 +1,532 @@
+                            if let Err(err) = io_ref.write_all(&res) {
+                                if debug {
+                                    eprintln!("IO write failed: {:?}", err);
+                                }
+                            }
+                        }
+                    } else if debug {
+                        eprintln!("SBP framing failed: {:?}", res.err());
+                    }
+                } else if debug {
+                    eprintln!("SBP parse failed: {:?}", sbp_msg.err());
+                }
+            } else if debug {
+                eprintln!("payload decode failed: {:?}", payload.err());
+            }
+        } else if debug {
+            eprintln!("no 'msg_type' field");
+        }
+    } else if debug {
+        eprintln!("no 'payload' field: {}", value);
+    }
+    Ok(())
+}
+
+/// The main SBP read loop, uses one buffer of size `BUF_SIZE`, and tracks the following
+///   markers for the buffer:
+///                                             
+///  ```text
+///
+///                                                         |<---remaining_length--->|
+///                                                         |                        |
+///                              |<-------read_length------>|                        |
+///                              |                          |                        |
+///                              read_offset                |                        |
+///         unconsumed_offset    ^                          |                        |
+///         ^                    |                          |                        |
+///         |                    |                          |                        |
+///  +------+--------------------+---------------------------------------------------+
+///  |######%%%%%%%%%%%%%%%%%%%%%@@@@@@@@@@@@@@@@@@@@@@@@@@@@&&&&&&&&&&&&&&&&&&&&&&&&|
+///  +-------------------------------------------------------------------------------+
+///
+///  ```
+///
+///  Where:
+///    ### -> old data, available to be dropped or overwritten
+///    %%% -> leftover data from previous read, not yet processed
+///    @@@ -> newly read data data, not yet processed
+///    &&& -> remaining free space in the buffer
+///
+///  The following definitions apply for the offsets / lengths tracked:
+///
+///  ```text
+///
+///    'unconsumed_offset' -> start of data that needs to be processed, could point to left over
+///      data from a previous read, or could be the same as read_offset.
+///
+///      Invariant: unconsumed_offset <= read_offset
+///
+///    'read_offset' -> the point where a new read should place its data
+///
+///    'read_length' -> the size of the last read, used to update `remaining_length`
+///
+///    'remaining_length' -> the remaining available space in the buffer
+///
+///  ```
+///
+pub fn sbp2json_read_loop(
```

Comment:
This seems to be handling the same kind of logic that I meant the `Parser` struct to in the main library. I don't think the `Parser` is particularly ergonomic as it current stands, and the fact that you had to implement something similar seems to confirm that.

Do you think we should reconsider the `Parser` object design? I think this is as good a time as any, and the longer we leave it as is makes it more difficult to  change it later.

Reply:
There are definitely things that could be more ergonomic, but the buffer logic here is a pretty verbatim copy of sbp2json in Python: https://github.com/swift-nav/libsbp/blob/master/python/sbp/sbp2json.py#L187 -- are you envisioning that most of this logic would live inside the parser object?

Reply:
Yeah, there's already some buffering logic in the parser object but it could be changed to mirror the python logic and the interface could probably made to be more rusty.